File: scripts/graveyard/86000.py
#!/usr/bin/env python3

import sys

if sys.version_info < (3, 0):
  print("Can't run on Python 2")
  sys.exit(-1)

unknown = []
known = []

fp = open("86000.txt", "r")

for line in fp:
  line = line.strip()
  if line == "": continue

  columns = line.split("\t")

  #print(str(columns))

  tokens =  columns[0].split()

  instruction = tokens[0].strip().lower()
  operands = ""

  if len(tokens) > 1:
    operands = "".join(tokens[1:])

  code = columns[1].split()

  if (len(columns) > 2):
    cycles = columns[2]
  else:
    cycles = "?"

  #print(cycles)

  #print(code)

  opcode = 0
  mask = 0
  skip = False

  for c in code[0]:
    if skip: skip = False; continue

    opcode = opcode << 1
    mask = mask << 1

    if c == "1": opcode |= 1
    if c in [ "0", "1" ]: mask |= 1
    if c in [ "b", "d", "i", "j" ]: skip = True

  #print(opcode)

  opcode = "0x%02x" % (opcode)
  mask = "0x%02x" % (mask)

  if operands == "":
    operands = "OP_NONE"
  elif operands == "r8":
    operands = "OP_RELATIVE8"
  elif operands == "@Ri":
    operands = "OP_AT_REG"
  elif operands == "d9":
    operands = "OP_ADDRESS"
  elif operands == "#i8":
    operands = "OP_IMMEDIATE"
  elif operands == "d9,b3":
    operands = "OP_ADDRESS_BIT"
  elif operands == "d9,r8":
    operands = "OP_ADDRESS_RELATIVE8"
  elif operands == "#i8,r8":
    operands = "OP_IMMEDIATE_RELATIVE8"
  elif operands == "@Rj,r8":
    operands = "OP_AT_REG_RELATIVE8"
  elif operands == "@Rj,#i8,r8":
    operands = "OP_AT_REG_IMMEDIATE_RELATIVE8"
  elif operands == "d9,b3,r8":
    operands = "OP_ADDRESS_BIT_RELATIVE8"
  elif operands == "r16":
    operands = "OP_RELATIVE16"
  elif operands == "a12":
    operands = "OP_ADDRESS12"
  elif operands == "a16":
    operands = "OP_ADDRESS16"
  elif operands == "@Ri,r8":
    operands = "OP_AT_REG_I_RELATIVE8"
  elif operands == "#i8,d9":
    operands = "OP_IMMEDIATE_ADDRESS"
  elif operands == "#i8,@Rj":
    operands = "OP_IMMEDIATE_AT_REG"
  else:
    if operands not in unknown: unknown.append(operands)

  print("  { \"" + instruction + "\", " + opcode + ", " + mask + ", " + operands + ", " + cycles + " },")

  if operands not in known: known.append(operands)

fp.close()

print(unknown)

for a in known:
  print(str(a) + ",")

File: disasm/86000.c
/**
 *  naken_asm assembler.
 *  Author: Michael Kohn
 *   Email: mike@mikekohn.net
 *     Web: http://www.mikekohn.net/
 * License: GPLv3
 *
 * Copyright 2010-2020 by Michael Kohn
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "disasm/86000.h"
#include "table/86000.h"

int get_cycle_count_86000(unsigned short int opcode)
{
  return -1;
}

int disasm_86000(
  struct _memory *memory,
  uint32_t address,
  char *instruction,
  int *cycles_min,
  int *cycles_max)
{
  int opcode, value, immediate, reg, bit;
  int8_t offset;
  int16_t offset16;
  int n;

  *cycles_min = -1;
  *cycles_max = -1;

  opcode = memory_read_m(memory, address);

  strcpy(instruction, "???");

  n = 0;
 
  while (table_86000[n].name != NULL)
  {
    if ((opcode & table_86000[n].mask) == table_86000[n].opcode)
    {
      *cycles_min = table_86000[n].cycles;
      *cycles_max = table_86000[n].cycles;

      switch (table_86000[n].type)
      {
        case OP_NONE:
          strcpy(instruction, table_86000[n].name);
          return 1;
        case OP_ADDRESS:
          value = ((opcode & 1) << 8) |memory_read_m(memory, address + 1);
          sprintf(instruction, "%s 0x%02x", table_86000[n].name, value);
          return 2;
        case OP_IMMEDIATE:
          immediate = memory_read_m(memory, address + 1);

          sprintf(instruction, "%s #0x%02x",
            table_86000[n].name,
            immediate);
          return 2;
        case OP_AT_REG:
          reg = opcode & 0x3;
          sprintf(instruction, "%s #0x%02x", table_86000[n].name, reg);
          return 1;
        case OP_ADDRESS_RELATIVE8:
          value = memory_read_m(memory, address + 1);
          offset = memory_read_m(memory, address + 2);

          sprintf(instruction, "%s 0x%02x, 0x%04x (offset=%d)",
            table_86000[n].name,
            value,
            address + 3 + offset,
            offset);
          return 3;
        case OP_IMMEDIATE_RELATIVE8:
          immediate = memory_read_m(memory, address + 1);
          offset = memory_read_m(memory, address + 2);

          sprintf(instruction, "%s #0x%02x, 0x%04x (offset=%d)",
            table_86000[n].name,
            immediate,
            address + 3 + offset,
            offset);
          return 3;
        case OP_AT_REG_IMMEDIATE_RELATIVE8:
          reg = opcode & 0x3;
          immediate = memory_read_m(memory, address + 1);
          offset = memory_read_m(memory, address + 2);

          sprintf(instruction, "%s @r%d, #0x%02x, %04x (offset=%d)",
            table_86000[n].name,
            reg,
            immediate,
            address + 3 + offset,
            offset);
          return 3;
        case OP_ADDRESS_BIT_RELATIVE8:
          bit = opcode & 0x7;
          value = ((opcode >> 4) & 1) | memory_read_m(memory, address + 1);
          offset = memory_read_m(memory, address + 2);

          sprintf(instruction, "%s #0x%02x, %d, 0x%04x (offset=%d)",
            table_86000[n].name,
            value,
            bit,
            address + 3 + offset,
            offset);
          return 3;
        case OP_RELATIVE8:
          offset = memory_read_m(memory, address + 1);

          sprintf(instruction, "%s 0x%04x (offset=%d)",
            table_86000[n].name,
            address + 2 + offset,
            offset);
          return 2;
        case OP_RELATIVE16:
          offset16 = (memory_read_m(memory, address + 2) << 8) |
                      memory_read_m(memory, address + 1);

          sprintf(instruction, "%s 0x%04x (offset=%d)",
            table_86000[n].name,
            address + 2 + offset16,
            offset16);
          return 3;
        case OP_ADDRESS12:
          value = ((opcode & 0x10) << 7) |
                  ((opcode & 0x07) << 8) |
                  memory_read_m(memory, address + 1);

          sprintf(instruction, "%s 0x%04x",
            table_86000[n].name,
            value);
          return 2;
        case OP_ADDRESS16:
          value = (memory_read_m(memory, address + 1) << 8) |
                   memory_read_m(memory, address + 2);

          sprintf(instruction, "%s 0x%04x",
            table_86000[n].name,
            value);
          return 3;
        case OP_ADDRESS_BIT:
          bit = opcode & 0x7;
          value = (((opcode >> 4) & 1) << 8) | memory_read_m(memory, address + 1);

          sprintf(instruction, "%s 0x%04x, %d",
            table_86000[n].name,
            value,
            bit);
          return 2;
        case OP_AT_REG_RELATIVE8:
          reg = opcode & 0x3;
          offset = memory_read_m(memory, address + 1);

          sprintf(instruction, "%s @r%d, 0x%04x (offset=%d)",
            table_86000[n].name,
            reg,
            address + 2 + offset,
            offset);
          return 2;
        case OP_IMMEDIATE_ADDRESS:
          value = ((opcode & 1) << 8) | memory_read_m(memory, address + 1);
          immediate = memory_read_m(memory, address + 2);

          sprintf(instruction, "%s #0x%02x, 0x%04x",
            table_86000[n].name,
            immediate,
            value);
          return 3;
        case OP_IMMEDIATE_AT_REG:
          reg = opcode & 0x3;
          immediate = memory_read_m(memory, address + 2);

          sprintf(instruction, "%s @r%d, 0x%04x",
            table_86000[n].name,
            immediate,
            address);
          return 2;
      }
    }

    n++;
  }

  return 1;
}

void list_output_86000(struct _asm_context *asm_context, uint32_t start, uint32_t end)
{
  int cycles_min = -1, cycles_max = -1, count;
  char instruction[128];
  char temp[32];
  char temp2[4];
  int n;

  fprintf(asm_context->list, "\n");

  while (start < end)
  {
    count = disasm_86000(&asm_context->memory, start, instruction, &cycles_min, &cycles_max);

    temp[0] = 0;

    for (n = 0; n < count; n++)
    {
      sprintf(temp2, "%02x ", memory_read_m(&asm_context->memory, start + n));
      strcat(temp, temp2);
    }

    fprintf(asm_context->list, "0x%04x: %-10s %-40s cycles: ", start, temp, instruction);

    if (cycles_min < 0)
    {
      fprintf(asm_context->list, "?");
    }
      else
    {
      fprintf(asm_context->list, "%d", cycles_min);
    }

    start += count;
  }
}

void disasm_range_86000(struct _memory *memory, uint32_t flags, uint32_t start, uint32_t end)
{
  char instruction[128];
  char temp[32];
  char temp2[4];
  int cycles_min = 0, cycles_max = 0;
  int count;
  int n;

  printf("\n");

  printf("%-7s %-5s %-40s Cycles\n", "Addr", "Opcode", "Instruction");
  printf("------- ------ ----------------------------------       ------\n");

  while (start <= end)
  {
    count = disasm_86000(memory, start, instruction, &cycles_min, &cycles_max);

    temp[0] = 0;

    for (n = 0; n < count; n++)
    {
      sprintf(temp2, "%02x ", memory_read_m(memory, start + n));
      strcat(temp, temp2);
    }

    printf("0x%04x: %-10s %-40s cycles: ", start, temp, instruction);

    if (cycles_min < 0)
    {
      printf("?");
    }
      else
    {
      printf("%d", cycles_min);
    }

    start += count;
  }
}

